include/aws/common/encoding.h

```diff
@@ -156,6 +164,16 @@ AWS_STATIC_IMPL bool aws_text_is_utf8(const uint8_t *bytes, size_t size);
  */
 AWS_COMMON_API bool aws_text_is_valid_utf8(struct aws_byte_cursor bytes);
 
+/**
+ * Scans every byte, and returns true if it is valid UTF8/ASCII as defined in RFC-3629 and
+ * user's customized validation callback. The function will return true for empty string
+ * regardless of the callback functions.
+ * The text does not need to begin with a UTF8 BOM.
+ */
+AWS_COMMON_API bool aws_text_is_valid_utf8_with_callback(
+    struct aws_byte_cursor bytes,
+    aws_utf8_customized_validator_callback_fn *validator_fn);
```

Comment:
Let's just make this a general purpose "decode" function, instead of a "validate" function.

`int aws_decode_utf8(struct aws_byte_cursor bytes, aws_on_utf8_codepoint_fn *on_codepoint, void *user_data);`

That makes more sense, people can use it to check length, or validate codepoints, or store codepoints... whatever they want!